src/Bicep.Core/TypeSystem/Az/AzResourceTypeProvider.cs

```diff
@@ -125,12 +126,32 @@ private static ObjectType SetBicepResourceProperties(ObjectType objectType, Reso
                 properties = properties.SetItem("resourceGroup", new TypeProperty("resourceGroup", LanguageConstants.String, TypePropertyFlags.DeployTimeConstant));
                 properties = properties.SetItem("subscriptionId", new TypeProperty("subscriptionId", LanguageConstants.String, TypePropertyFlags.DeployTimeConstant));
             }
+
+            var functions = GetTypeBicepFunctions(objectType.Name);
+
             return new NamedObjectType(
                 objectType.Name,
                 objectType.ValidationFlags,
                 isExistingResource ? ConvertToReadOnly(properties.Values) : properties.Values,
                 objectType.AdditionalPropertiesType,
-                isExistingResource ? ConvertToReadOnly(objectType.AdditionalPropertiesFlags) : objectType.AdditionalPropertiesFlags);
+                isExistingResource ? ConvertToReadOnly(objectType.AdditionalPropertiesFlags) : objectType.AdditionalPropertiesFlags,
+                functions);
+        }
+
+        private static IEnumerable<FunctionOverload> GetTypeBicepFunctions(string name)
+        {
+            switch (name.ToLowerInvariant())
+            {
+                case "microsoft.keyvault/vaults":
+                    yield return new FunctionOverloadBuilder("getSecret")
+                        .WithReturnType(LanguageConstants.KeyVaultSecretReference)
```

Comment:
Why does this return a `KeyVaultSecretReference` type (extension of object) rather than a secure string with `AllowKeyVaultSecretReferenceAssignment` flag?

Reply:
The main reason behind this (and introducing a type) was reason, that I wasn't able to pin-point allowed location to exactly the parameter of type string with secure decorator. The closest I could get was to detect that we're in an object property which is params inside module.

Then, if I'd use only string with ValidationFlag only, I was getting errors like: `cannot assign type string to type string`. A
Of course I could add another cases and additional diagnostic errors to handle this error, but in my opinion this is even worse (branching new "special" cases in various, distinct places) than introducing type and taking advantage of the type checking mechanism.

Reply:
@anthony-c-martin @majastrz here is where we ensure that function is only available for using on keyVault type.